Move restart restore further up in initial setup

loganharbour opened the pull request:
Closes #25249

loganharbour committed:
Move MaterialData into MaterialPropertyStorage for simplicity

loganharbour committed:
Tighten registry APIs, add optional getter

loganharbour committed:
Fix typos

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1603027/) on d3128a8 wanted to post the following:

The following commit(s) contain fixups:

- d3128a8e70: fixup! 1704259010d19dbacab289dc56c7fe821f314631

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1603102/) on ec565f0 wanted to post the following:

The following commit(s) contain fixups:

- ec565f00a1: fixup! 1704259010d19dbacab289dc56c7fe821f314631
- 5fee597449: fixup! 1704259010d19dbacab289dc56c7fe821f314631
- d3128a8e70: fixup! 1704259010d19dbacab289dc56c7fe821f314631

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1603143/) on ec565f0 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>88aedd</th>
      <th colspan=4>#24762 ec565f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.40%</td>
      <td>85.40%</td>
      <td>+0.00%</td>
      <td>82.72%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>88808</td>
      <td>88786</td>
      <td>-22</td>
      <td>134</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15179</td>
      <td>15174</td>
      <td>-5</td>
      <td>28</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24762/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24762/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>88aedd</th>
      <th colspan=4>#24762 ec565f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.85%</td>
      <td>83.85%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>10827</td>
      <td>10827</td>
      <td>-</td>
      <td>2</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2085</td>
      <td>2085</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24762/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24762/coverage/navier_stokes)

### Porous flow

<table>
  <thead>
    <tr>
      <th></th>
      <th>88aedd</th>
      <th colspan=4>#24762 ec565f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>95.69%</td>
      <td>95.69%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>10778</td>
      <td>10780</td>
      <td>+2</td>
      <td>6</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>485</td>
      <td>485</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24762/coverage_diff/porous_flow)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24762/coverage/porous_flow)

### Xfem

<table>
  <thead>
    <tr>
      <th></th>
      <th>88aedd</th>
      <th colspan=4>#24762 ec565f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>81.95%</td>
      <td>81.95%</td>
      <td>-</td>
      <td>80.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>6555</td>
      <td>6555</td>
      <td>-</td>
      <td>12</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1444</td>
      <td>1444</td>
      <td>-</td>
      <td>3</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24762/coverage_diff/xfem)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24762/coverage/xfem)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24762/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24762/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24762/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24762/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24762/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24762/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24762/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24762/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24762/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24762/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24762/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24762/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24762/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24762/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24762/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24762/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24762/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24762/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24762/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24762/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24762/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24762/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24762/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24762/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24762/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24762/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24762/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24762/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 82.72% is less than the suggested 90.0%
- `xfem` new line coverage rate 80.00% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1603125/) on ec565f0 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24762/site)

This comment will be updated on new commits.

loganharbour committed:
Add namespace for unity builds

loganharbour committed:
Remove unused property name

loganharbour committed:
Add extra checking for types

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1611115/) on 09f2413 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24762/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24762/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c21cec693e04b57dcc8e61efc9348e157350b175`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1611115/) on 09f2413 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24762/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24762/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c21cec693e04b57dcc8e61efc9348e157350b175`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1611115/) on 09f2413 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24762/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24762/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c21cec693e04b57dcc8e61efc9348e157350b175`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1611115/) on 09f2413 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24762/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24762/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c21cec693e04b57dcc8e61efc9348e157350b175`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1611115/) on 09f2413 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24762/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24762/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c21cec693e04b57dcc8e61efc9348e157350b175`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1611207/) on 2e30b5a wanted to post the following:

The following commit(s) contain fixups:

- 2e30b5ab0f: fixup! 09f24133344177500bb2fee1a9ec3925ceae458f

loganharbour committed:
Add an extra clear :(

loganharbour committed:
Reserve space while we're here

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1611592/) on aa90709 wanted to post the following:

The following commit(s) contain fixups:

- aa90709487: fixup! 69b6f5d0fb4172b27bcbf9e78368c6cce1001d01
- 2e30b5ab0f: fixup! 09f24133344177500bb2fee1a9ec3925ceae458f

loganharbour committed:
Add myself as a codeowner to materials and restart stuff

loganharbour committed:
Reduce code duplication

loganharbour committed:
Add mat prop ID to data, reduce interface duplication

loganharbour committed:
Add the ability to skip data

loganharbour committed:
Begin work on moving stateful property load further up in initial setup

refs #25249

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1618741/) on ba70e32 wanted to post the following:

The following commit(s) contain fixups:

- 037c16851a: fixup! c37781e645b1f07e5598ece2e29060eb7be06372

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1619290/) on 252f7d8 wanted to post the following:

The following commit(s) contain fixups:

- 252f7d8c17: fixup! ba70e32b54e1d82e985e5d30e86247c6aa768cc3
- 037c16851a: fixup! c37781e645b1f07e5598ece2e29060eb7be06372

loganharbour committed:
Make all the things final that we can

refs #25249

loganharbour committed:
Add some constexpr, add unitialized value

refs #25249

loganharbour committed:
Use the clone instead of additional logic

refs #25249

loganharbour committed:
Fix sizing logic

refs #25249

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1622587/) on 6bafd8a wanted to post the following:

The following commit(s) contain fixups:

- 6bafd8a0b0: fixup! 82f97b549b6ed96dfd1175063f29263293a5bf7d
- 2ad5be46b4: fixup! 82f97b549b6ed96dfd1175063f29263293a5bf7d
- 252f7d8c17: fixup! ba70e32b54e1d82e985e5d30e86247c6aa768cc3
- 037c16851a: fixup! c37781e645b1f07e5598ece2e29060eb7be06372

loganharbour committed:
Cast to avoid comparison error

refs #25249

loganharbour committed:
Simplify resize logic, add comments

refs #25249

loganharbour committed:
Regold file

refs #25249
This actually improves the behavior of this input; instead of
having an incorrect penetration initially, we now have a
physically correct one

loganharbour committed:
Request all of this up front

refs #25249

loganharbour committed:
Add ability to declare reporters late

We need this because we need them to be delcared earlier
than initialSetup() when the restartable reload is moved
back more

refs #25249

loganharbour committed:
Fix true initial counts, cleanup map insertion

refs #25249

loganharbour committed:
Add missing header

refs #25249

loganharbour committed:
Correct method in docs

refs #25249

loganharbour committed:
Make the stateful prop loop thread consistent

refs #25249

loganharbour committed:
Add missing overrides

refs #25249

loganharbour committed:
Improve dataLoad errors for stateful properties

refs #25249

loganharbour committed:
Add error checking for a stateful material property mismatch

refs #25249

loganharbour commented:
@lindsayad this is ready for review.

Remaining

- [x] bison: https://github.inl.gov/ncrc/bison/pull/5567
- [ ] bison: https://github.inl.gov/ncrc/bison/pull/5561, should be merged soon
- [ ] griffin: https://github.inl.gov/ncrc/griffin/pull/1355, breaking change ready
- [ ] zapdos: need to create

loganharbour committed:
Allow an invalid default

refs #25249

loganharbour committed:
Cleanup multiapp restore and add comments about its location

Also removes the now-unnecessary reprojection step

refs #25249

loganharbour committed:
Unify special material property IDs

refs #25249

loganharbour commented:
Sorry all for the spam. I am actually done with this now.

lindsayad commented on framework/src/materials/MaterialPropertyStorage.C:
```diff
@@ -361,30 +366,191 @@ MaterialPropertyStorage::initProps(MaterialData & material_data,
+    std::size_t num_elems = storage.setProps(state).size();
+    dataStore(stream, num_elems, nullptr);
+
+    for (auto & elem_side_map_pair : storage.setProps(state))
```
```suggestion
    for (const auto & [elem, side_map] : storage.setProps(state))
```

loganharbour replied:
Can't due to non-cost `dataStore(std::istream &, T & data, ...)`

lindsayad commented on framework/src/materials/MaterialPropertyStorage.C:
```diff
@@ -361,30 +366,191 @@ MaterialPropertyStorage::initProps(MaterialData & material_data,
+
+    for (auto & elem_side_map_pair : storage.setProps(state))
+    {
+      const Elem * elem = elem_side_map_pair.first;
```
```suggestion
```

lindsayad commented on framework/src/materials/MaterialPropertyStorage.C:
```diff
@@ -361,30 +366,191 @@ MaterialPropertyStorage::initProps(MaterialData & material_data,
+      mooseAssert(elem, "Null element");
+      dataStore(stream, elem, context);
+
+      auto & side_map = elem_side_map_pair.second;
```
```suggestion
```

lindsayad commented on framework/src/materials/MaterialPropertyStorage.C:
```diff
@@ -361,30 +366,191 @@ MaterialPropertyStorage::initProps(MaterialData & material_data,
+      std::size_t num_sides = side_map.size();
+      dataStore(stream, num_sides, nullptr);
+
+      for (auto & side_props_pair : side_map)
+      {
+        unsigned int side = side_props_pair.first;
+        dataStore(stream, side, nullptr);
+
+        auto & props = side_props_pair.second;
```
```suggestion
      for (const auto & [side, props] : side_map)
      {
        dataStore(stream, side, nullptr);

```

lindsayad commented on test/tests/materials/stateful_prop/tests:
```diff
@@ -156,4 +156,31 @@
+                  Outputs/checkpoint=true
+                  Outputs/file_base=restart_different_props'
+      recover = false
+      detail = 'storing the named stateful material properties during checkpoint'
```
```suggestion
      detail = 'storing the named stateful material properties during checkpoint, and'
```

loganharbour replied:
These are so wildly inconsistent (just above)

```
    requirement = 'The system shall be able to correctly handle stateful materials'
    [reg]
      type = 'Exodiff'
      input = 'stateful_reg.i'
      exodiff = 'stateful_reg_out.e'
      detail = 'when using a regular material property.'
    []
    [ad]
      type = 'Exodiff'
      input = 'stateful_ad.i'
      exodiff = 'stateful_ad_out.e'
      detail = 'when using an AD material property.'
    []
    [template_reg]
      type = 'Exodiff'
    ...
```

lindsayad commented on test/tests/materials/stateful_prop/tests:
```diff
@@ -156,4 +156,31 @@
+      recover = false
+      prereq = 'restart_different_props/store'
+      expect_err = 'dummy_prop.*dummy_prop.*dummy_prop_2'
+      detail = 'loading the stateful material properties during recover and comparing to the declared stateful material properties'
```
```suggestion
      detail = 'loading the stateful material properties during recover and comparing to the declared stateful material properties.'
```

lindsayad reviewed (changes requested):
just small things

loganharbour committed:
Make data const

loganharbour commented on framework/include/interfaces/XFEMInterface.h:
```diff
@@ -139,8 +134,8 @@ class XFEMInterface : public ConsoleStreamInterface
   FEProblemBase * _fe_problem;
-  std::vector<std::shared_ptr<MaterialData>> * _material_data;
-  std::vector<std::shared_ptr<MaterialData>> * _bnd_material_data;
+  std::vector<MaterialData *> _material_data;
+  std::vector<MaterialData *> _bnd_material_data;
```
just the number of threads

loganharbour commented on framework/include/materials/MaterialData.h:
```diff
@@ -31,7 +31,10 @@ class XFEM;
-  MaterialData(MaterialPropertyStorage & storage);
+  MaterialData(MaterialPropertyStorage & storage, const THREAD_ID tid);
+
+  /// The max time state supported (2 = older)
+  static constexpr unsigned int max_state = 2;
```
correct, used in `MaterialPropertyStorage`

loganharbour committed:
Remove default on parameters

refs #25249

loganharbour commented on framework/include/materials/MaterialPropertyInterface.h:
```diff
@@ -453,6 +453,17 @@ class MaterialPropertyInterface
   /// Current threaded it
   const THREAD_ID _mi_tid;
 
+private:
+  /// BoundaryRestricted flag
+  const bool _mi_boundary_restricted;
+
+protected:
```
It needs to be like this; the member below (which is not private) depends on it:

```
    _mi_boundary_restricted(!boundary_ids.empty() &&
                            BoundaryRestrictable::restricted(boundary_ids)),
    _material_data_type(getMaterialDataType()),
```

with
```
_material_data_type

lindsayad replied:
You could also make `_mi_boundary_restricted` protected. If `_material_data_type` is already not encapsulated, I don't see much harm in its precursor being in the same state

lindsayad replied:
I created lindsayad/moose@0956fe4075 which removes the access level flip while keeping `_mi_boundary_restricted` `private`. Unfortunately github doesn't give me the ability to make a pull request into your fork

loganharbour commented on framework/src/base/Moose.C:
```diff
@@ -334,6 +335,7 @@ addActionTypes(Syntax & syntax)
                            "(add_vector_postprocessor)" // MaterialVectorPostprocessor requires this
                                                         // to be after material objects are created.
                            "(add_reporter)"
+                           "(declare_late_reporters)"
```
It's used exclusively for this: https://github.com/loganharbour/moose/blob/cc58ca0d5bcd705a13e196f1fa264b7c46d0b161/framework/src/reporters/AccumulateReporter.C#L33

which allows for lazy typed accumulation

loganharbour committed:
Remove newline

loganharbour committed:
Remove unnecessary ref

loganharbour commented on framework/src/materials/MaterialData.C:
```diff
@@ -67,3 +68,27 @@ MaterialData::mooseErrorHelper(const MooseObject & object, const std::string_vie
   object.mooseError(error);
 }
+
+bool
+MaterialData::hasProperty(const std::string & prop_name) const
+{
+  return _storage.hasProperty(prop_name);
+}
+
+unsigned int
+MaterialData::getPropertyId(const std::string & prop_name) const
+{
+  return _storage.getMaterialPropertyRegistry().getID(prop_name);
+}
+
+void
+MaterialData::eraseProperty(const Elem * elem)
+{
+  _storage.eraseProperty(elem);
+}
+
+unsigned int
+MaterialData::addPropertyHelper(const std::string & prop_name, const unsigned int state)
+{
+  return _storage.addProperty(prop_name, state);
+}
```
Can't due to header includes

loganharbour commented on framework/src/materials/MaterialPropertyStorage.C:
```diff
@@ -361,30 +366,191 @@ MaterialPropertyStorage::initProps(MaterialData & material_data,
 {
-  unsigned int max_state = storage.maxState();
-  dataStore(stream, max_state, context);
+  auto num_states = storage.numStates();
```
Can't due to non-cost `dataStore(std::istream &, T & data, ...)`

lindsayad replied:
That's not an issue unless `data` calls a non-const method. `T` should deduce the cv qualifications

lindsayad replied:
This program for example works fine
```c++
template <typename T>
void
foo(T &)
{
}

int
main()
{
  const int i = 0;
  double j = 0;
  foo(i);
  foo(j);
}
```

loganharbour commented on framework/src/materials/MaterialPropertyStorage.C:
```diff
@@ -361,30 +366,191 @@ MaterialPropertyStorage::initProps(MaterialData & material_data,
+
+  // Store the material property ID -> name map for mapping back
+  const auto & registry = storage.getMaterialPropertyRegistry();
+  std::vector<std::string> ids_to_names(registry.idsToNamesBegin(), registry.idsToNamesEnd());
```
Can't due to non-cost `dataStore(std::istream &, T & data, ...)`

lindsayad replied:
same as above

loganharbour replied:
```
In file included from /ssd/tmp/harblh/worktrees/moose/material_rewrite_2/framework/build/header_symlinks/MaterialProperty.h:19,
                 from /ssd/tmp/harblh/worktrees/moose/material_rewrite_2/framework/build/header_symlinks/MaterialData.h:12,
                 from /ssd/tmp/harblh/worktrees/moose/material_rewrite_2/framework/build/header_symlinks/BlockRestrictable.h:15,
                 from /ssd/tmp/harblh/worktrees/moose/material_rewrite_2/framework/build/header_symlinks/MaterialBase.h:14,
                 from /ssd/tmp/harblh/worktrees/moose/material_rewrite_2/framework/build/header_symlinks/Material.h:13,
                 from /ssd/tmp/harblh/worktrees/moose/material_rewrite_2/framework/build/header_symlinks/ADPiecewiseLinearInterpolationMaterial.h:12,
                 from /ssd/tmp/harblh/worktrees/moose/material_rewrite_2/framework/src/materials/ADPiecewiseLinearInterpolationMaterial.C:10,
                 from /ssd/tmp/harblh/worktrees/moose/material_rewrite_2/framework/build/unity_src/materials_Unity.C:2:
/ssd/tmp/harblh/worktrees/moose/material_rewrite_2/framework/build/header_symlinks/DataIO.h: In instantiation of 'void dataStore(std::ostream&, T&, void*) [with T = const std::vector<std::__cxx11::basic_string<char> >; std::ostream = std::basic_ostream<char>]':
/ssd/tmp/harblh/worktrees/moose/material_rewrite_2/framework/src/materials/MaterialPropertyStorage.C:382:42:   required from here
/ssd/tmp/harblh/worktrees/moose/material_rewrite_2/framework/build/header_symlinks/DataIO.h:165:17: error: static assertion failed: Cannot serialize a class that is not trivially copyable!
Write a custom dataStore() template specialization!


   static_assert(std::is_trivially_copyable<T>::value,
                 ^~~
make: *** [/ssd/tmp/harblh/worktrees/moose/material_rewrite_2/framework/build.mk:145: /ssd/tmp/harblh/worktrees/moose/material_rewrite_2/framework/build/unity_src/materials_Unity.x86_64-pc-linux-gnu.opt.lo] Error 1
```

We don't instantiate `store/loadHelper(..., const std::vector<T> ...)`

lindsayad replied:
K we don't have to do it now, but I think all `dataStores` should be `const`

loganharbour replied:
@roystgnr and I have both tried to go down this path... to frustration and not enough time

roystgnr replied:
Yeah, @lindsayad; IMHO it's obvious that all dataStores should be const, and it *should* be an easy fix ... but I tried to fix it and everything broke, and before me Logan tried to fix it and everything broke, and before him @permcody tried to fix it and everything broke.  There were some [dire warnings on Slack](https://moosedevelopers.slack.com/archives/C01054VRUEM/p1675802764882089).

You know, with that track record, it would have been funny to have made you fix it before promoting you to MOOSE lead.  Can someone really be the Once and Future King of MOOSE unless they're the only one who can pull the Sword of Const-Correctness from the Stone of Legacy Code?

lindsayad replied:
We can add someone else's name as CODEOWNER to all the framework files and see whether they deem it a promotion 😆 

It would be better to have a concrete documented reason why it fails, then to just vaguely wish other developers good luck

loganharbour replied:
...is this resolved?

lindsayad replied:
> K we don't have to do it now,

This is what I said. As there was another unresolved conversation and there was ongoing discussion on this thread, I did not immediately mark it as resolved. I think I give very punctual reviews and even submit patches when my review has not been addressed. I don't really want to be hastened further

loganharbour committed:
Do some grammar

loganharbour committed:
Add a const

refs #25249

loganharbour committed:
Use structured bindings

lindsayad committed:
Limit changes in access level in MaterialPropertyInterface
